fabtests/benchmarks/rdm_bw_mt.c

```diff
@@ -0,0 +1,767 @@
+	if (ret) {
+		printf("send read_cq error: %d\n", ret);
+		return ret;
+	}
+	return 0;
+}
+
+static int bw_recv(void *context)
+{
+	int ret = FI_SUCCESS;
+	struct thread_args *targs = context;
+
+	ret = post_recv(context);
+	if (ret)
+		return ret;
+
+	ret = read_cq(targs->cq);
+	if (ret) {
+		printf("recv read_cq error: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void *uni_bandwidth(void *context)
+{
+	int i, ret;
+
+	pthread_barrier_wait(&barrier);
+	for (i = 0; i < opts.warmup_iterations; i++) {
+		ret = opts.dst_addr ? bw_send(context) : bw_recv(context);
+		if (ret) {
+			((struct thread_args *) context)->ret = ret;
+			printf("ep[%d] warmup failed iter %d\n", targs->id, i);
+			break;
+		}
+	}
+
+	pthread_barrier_wait(&barrier);
+	if (targs->id == 0)
+		ft_start();
+	for (i = 0; i < opts.iterations; i++) {
+		ret = opts.dst_addr ? bw_send(context) : bw_recv(context);
+		if (ret) {
+			((struct thread_args *) context)->ret = ret;
+			break;
+		}
+	}
+	pthread_barrier_wait(&barrier);
+	if (targs->id == 0)
+		ft_stop();
+
+	return  NULL;
+}
+
+static void *bi_bandwidth(void *context)
+{
+	int i, ret;
+	struct thread_args *targs = context;
+
+	pthread_barrier_wait(&barrier);
+	for (i = 0; i < opts.warmup_iterations; i++) {
+		ret = opts.dst_addr ? bw_send(context) : bw_recv(context);
+		if (ret) {
+			((struct thread_args *) context)->ret = ret;
+			printf("%d warmup failed iter %d\n", targs->id, i);
+			break;
+		}
+		ret = opts.dst_addr ? bw_recv(context) : bw_send(context);
+		if (ret) {
+			((struct thread_args *) context)->ret = ret;
+			printf("%d warmup failed iter %d\n", targs->id, i);
+			break;
+		}
+	}
+
+	pthread_barrier_wait(&barrier);
+	if (targs->id == 0)
+		ft_start();
+
+	for (i = 0; i < opts.iterations; i++) {
+		ret = opts.dst_addr ? bw_send(context) : bw_recv(context);
+		if (ret) {
+			((struct thread_args *) context)->ret = ret;
+			printf("%d warmup failed iter %d\n", targs->id, i);
+			break;
+		}
+		ret = opts.dst_addr ? bw_recv(context) : bw_send(context);
+		if (ret) {
+			((struct thread_args *) context)->ret = ret;
+			printf("%d warmup failed iter %d\n", targs->id, i);
+			break;
+		}
+	}
+	pthread_barrier_wait(&barrier);
+	if (targs->id == 0)
+		ft_stop();
+
+	return NULL;
+}
+
+static int run_size(size_t size)
+{
+	int ret = FI_SUCCESS;
+	int err = FI_SUCCESS;
+	int i;
+
+	for (i = 0; i < num_eps; i++) {
+		targs[i].size = size;
+		targs[i].id = i;
+		targs[i].ret = FI_SUCCESS;
+		ret = reg_mrs(&targs[i]);
+		if (ret)
+			goto no_mem;
+	}
+
+	for (i = 0; i < num_eps; i++) {
+		ret = pthread_create(&targs[i].thread, NULL,
+				     bidir ? bi_bandwidth : uni_bandwidth,
+				     &targs[i]);
+		if (ret) {
+			printf("pthread_create failed: %d\n", ret);
+			return ret;
+		}
+	}
+	for (i = 0; i < num_eps; i++) {
+		pthread_join(targs[i].thread, NULL);
+		if (targs[i].ret) {
+			ret = targs[i].ret;
+			goto no_mem;
+		}
+	}
+
+	show_perf(NULL, size, opts.iterations, &start, &end, num_eps);
+
+no_mem:
```

Comment:
This happens on test failure, memory failure, and success, just rename to out or free